x/mint/keeper/emissions.go

```diff
@@ -0,0 +1,131 @@
+	return math.ZeroInt()
+}
+
+// helper function to get the number of staked tokens on the network
+// includes both tokens staked by cosmos validators (cosmos staking)
+// and tokens staked by reputers (allora staking)
+func GetNumStakedTokens(ctx context.Context, k Keeper) (math.Int, error) {
+	cosmosValidatorsStaked, err := k.StakingTokenSupply(ctx)
+	if err != nil {
+		return math.Int{}, err
+	}
+	reputersStaked, err := k.emissionsKeeper.GetTotalStake(ctx)
+	if err != nil {
+		return math.Int{}, err
+	}
+	return cosmosValidatorsStaked.Add(math.NewIntFromBigInt(reputersStaked.BigInt())), nil
+}
+
+// The total amount of tokens emitted as rewards at timestep i
+// E_i = e_i*N_{staked,i}
+// where e_i is the emission per unit staked token
+// and N_{staked,i} is the total amount of tokens staked at timestep i
+// THIS FUNCTION TRUNCATES THE RESULT DIVISION TO AN INTEGER
+func GetTotalEmissionPerTimestep(
+	rewardEmissionPerUnitStakedToken math.LegacyDec,
+	numStakedTokens math.Int,
+) math.Int {
+	return rewardEmissionPerUnitStakedToken.MulInt(numStakedTokens).TruncateInt()
+}
+
+// Target Reward Emission Per Unit Staked Token
+// controls the inflation rate of the token issuance
+//
+// ^e_i = ((f_e*T_{total,i}) / N_{staked,i}) * (N_{circ,i} / N_{total,i})
+//
+// where T_{total,i} is the total number of tokens held by the ecosystem
+// treasury, N_{total,i} is the total token supply, N_{circ,i} is the
+// circulating supply, and N_{staked,i} is the staked supply. The
+// factor f_e = 0.015 month^{−1} represents the fraction of the
+// ecosystem treasury that would ideally be emitted per unit time.
+// pass f_e as a fractional value, numerator and denominator as separate args
+func GetTargetRewardEmissionPerUnitStakedToken(
+	fEmissionNumerator math.Int,
+	fEmissionDenominator math.Int,
+	ecosystemBalance math.Int,
+	networkStaked math.Int,
+	circulatingSupply math.Int,
+	totalSupply math.Int,
+) (math.LegacyDec, error) {
+	if fEmissionDenominator.IsZero() ||
+		networkStaked.IsZero() ||
+		totalSupply.IsZero() {
+		return math.LegacyDec{}, errors.Wrapf(
+			types.ErrZeroDenominator,
+			"denominator is zero: %s | %s",
+			networkStaked.String(),
+			fEmissionDenominator.String(),
+		)
+	}
+	// T_{total,i} = ecosystemBalance
+	// N_{staked,i} = networkStaked
+	// N_{circ,i} = circulatingSupply
+	// N_{total,i} = totalSupply
+	ratioCirculating := circulatingSupply.ToLegacyDec().Quo(totalSupply.ToLegacyDec())
+	ratioEcosystemToStaked := ecosystemBalance.ToLegacyDec().Quo(networkStaked.ToLegacyDec())
+	ret := fEmissionNumerator.ToLegacyDec().
+		Mul(ratioEcosystemToStaked).
+		Mul(ratioCirculating).
+		Quo(fEmissionDenominator.ToLegacyDec())
+	if ret.IsNegative() {
+		return math.LegacyDec{}, errors.Wrapf(
+			types.ErrNegativeTargetEmissionPerToken,
+			"target emission per token is negative: %s | %s | %s",
+			ratioCirculating.String(),
+			ratioEcosystemToStaked.String(),
+			ret.String(),
+		)
+	}
+
+	return ret, nil
+}
```

Comment:
its insane to me how much code there is for essentially
```
if (C == 0 || E == 0 || F == 0) throw 'division by zero'
result = (A * (B / C) * (D / E)) / F
if (result < 0) throw 'negative emissions'
return result
```

Reply:
is there any way to simplify this code? or is go just super crufty

Reply:
I'd probably rewrite it like, biggest goal of idiomatic golang is to reduce mental unrolling when we come back to it in 6 months.

```go
// CalculateTargetRewardEmission calculates the target reward emission per unit of staked token.
func CalculateTargetRewardEmission(
	emissionNumerator, emissionDenominator, ecosystemBalance,
	networkStaked, circulatingSupply, totalSupply math.Int,
) (math.LegacyDec, error) {
	if emissionDenominator.IsZero() || networkStaked.IsZero() || totalSupply.IsZero() {
		return math.LegacyDec{}, fmt.Errorf(
			"invalid input: denominator, network staked, or total supply is zero",
		)
	}

	// Calculate the ratios needed for the target emission calculation
	circulatingRatio := circulatingSupply.ToLegacyDec().Quo(totalSupply.ToLegacyDec())
	ecosystemStakedRatio := ecosystemBalance.ToLegacyDec().Quo(networkStaked.ToLegacyDec())

	// Calculate the target emission per unit of staked token
	targetEmission := emissionNumerator.ToLegacyDec().
		Mul(ecosystemStakedRatio).
		Mul(circulatingRatio).
		Quo(emissionDenominator.ToLegacyDec())

	if targetEmission.IsNegative() {
		return math.LegacyDec{}, fmt.Errorf(
			"negative target emission: circulating ratio %s, ecosystem to staked ratio %s, emission %s",
			circulatingRatio.String(), ecosystemStakedRatio.String(), targetEmission.String(),
		)
	}

	return targetEmission, nil
}
```

Reply:
yeah go is extremely verbose it sucks